src/librustc_mir/interpret/eval_context.rs

```diff
@@ -576,18 +576,22 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tc
         Ok(())
     }
 
-    pub fn const_eval(&self, gid: GlobalId<'tcx>) -> EvalResult<'tcx, &'tcx ty::Const<'tcx>> {
+    pub fn const_eval_raw(
+        &self,
+        gid: GlobalId<'tcx>,
+    ) -> EvalResult<'tcx, MPlaceTy<'tcx, M::PointerTag>> {
         let param_env = if self.tcx.is_static(gid.instance.def_id()).is_some() {
             ty::ParamEnv::reveal_all()
         } else {
             self.param_env
         };
-        self.tcx.const_eval(param_env.and(gid)).map_err(|err| {
+        let val = self.tcx.const_eval_raw(param_env.and(gid)).map_err(|err| {
```

Comment:
Hm, on second thought I have doubts about using the raw query here. Couldn't that mean that unvalidated values flow into promotion?

However, memory *already* used the raw query when following a pointer to a static, so this is not a new concern.

Reply:
As long as the collector and `codegen` keep using `const_eval`, everything's fine. Maybe add a comment about that.

Reply:
you can already have this today if you write `FOO` as a `const fn`. I don't think we can do much there.

Reply:
`const fn` have more strict static checks though, for that precise reason (e.g., no `unsafe`).

Reply:
generic consts should totally get those restrictions, too, imo.

Also we're likely to start lifting these restrictions some day

Reply:
To counter my own argument: We will eventually allow more `unsafe` in `const fn` though, we just have to figure out what to do about `unconst`. Eventually we will allow the same things in `const` and `const fn`, I hope.

Moreover, the result of this computation, before flowing into run-time code, *will* be validated.

This leaves one concern, and that is promoteds: We promote all uses of `const`, but not all `const fn`. But I guess when evaluating a promoted, whether we fail here because a dependency is invalid or fail later, in both cases we fail and abort compilation, so it doesn't really make such a big difference.